File: rust/404.sum-of-left-leaves.rs
/*
 * @lc app=leetcode id=404 lang=rust
 *
 * [404] Sum of Left Leaves
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn sum_of_left_leaves(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut sum = 0;
        if let Some(node) = root {
            let nb = node.borrow();
            let left_node = &nb.left;
            if let Some(ln) = left_node {
                let lnb = ln.borrow();
                if lnb.left.is_none() && lnb.right.is_none() {
                    sum += lnb.val;
                } else {
                    sum += Self::sum_of_left_leaves(left_node.clone());
                }
            }
            sum += Self::sum_of_left_leaves(nb.right.clone());
        }
        return sum;
    }
}
// @lc code=end

File: javascript/404.sum-of-left-leaves.js
/*
 * @lc app=leetcode id=404 lang=javascript
 *
 * [404] Sum of Left Leaves
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumOfLeftLeaves = function (root) {
  // Solution 1: recursion
  // let sum = 0;
  // if (root) {
  //   const { left, right } = root;
  //   if (left) {
  //     if (!left.left && !left.right) sum += left.val;
  //     else sum += sumOfLeftLeaves(left);
  //   }
  //   sum += sumOfLeftLeaves(right);
  // }
  // return sum;

  // Solution 2: DFS
  function DFS(node, isLeft = false) {
    if (node) {
      if (isLeft && !node.left && !node.right) sum += node.val;
      else {
        DFS(node.left, true);
        DFS(node.right);
      }
    }
  }
  let sum = 0;
  DFS(root);
  return sum;
};
// @lc code=end
